### PR TITLE
net: l2: ieee802154: improved test coverage

### PR DESCRIPTION
**Pull request by ghost**
This PR is the first of several PRs that will develop the IEEE 802.15.4 stack in direction of TSCH support, see the corresponding [RFC](https://github.com/zephyrproject-rtos/zephyr/issues/50336) that lays out the overall implementation approach.

Supporting IEEE 802.15.4-2020 features in general and TSCH in particular requires substantial changes to the stack in order to align it with IEEE 802.15.4-2020 concepts, especially in the area of MHR parsing, support of arbitrary optional information elements (IEs), standards confirming configuration of MAC/PHY PIB attributes and incoming and outgoing security procedures.

The internal API of the IEEE 802.15.4 stack will have to be made more extensible and configurable to support these features. A preview of further changes to come is given in the [open development branch](https://github.com/zephyrproject-rtos/zephyr/compare/main...fgrandel:zephyr:feat/ieee802154-tsch-ie-support) to motivate why such preparative work is required.

**To ensure that none of these upcoming changes will break backwards compatibility, broad test coverage is introduced in this first PR.**

### CODE REVIEW

**Comment by fgrandel** [comment]
@nandojve Heads up to you as you might review this PR as well.

File: tests/net/ieee802154/l2/src/ieee802154_test.c [hunk 1]
```diff
@@ -983,6 +987,7 @@ ZTEST(ieee802154_l2, test_parsing_ack_pkt)
 	zassert_true(ret, "ACK parsed");
 }
 
+#ifdef CONFIG_NET_SOCKETS
```
Comment (rlubos) [inline tests/net/ieee802154/l2/src/ieee802154_test.c:1144]: I wonder whether it'd make sense to define a separate `ZTEST_SUITE()` for raw socket tests?

**Review by rlubos — COMMENTED** [review]
Just some nits, but overall it looks quite reasonable to me.

File: tests/net/ieee802154/l2/src/ieee802154_test.c [hunk 2]
```diff
@@ -177,6 +177,42 @@ static int tear_down_short_addr(struct net_if *iface, struct ieee802154_context
 	return ret;
 }
 
+static bool set_up_security(uint8_t security_level)
+{
+	struct ieee802154_security_params params;
+
+	if (security_level == IEEE802154_SECURITY_LEVEL_NONE)
+		return true;
```
[the remark this answers is not in the document]
Comment (fgrandel) [reply tests/net/ieee802154/l2/src/ieee802154_test.c]: Fixed

File: tests/net/ieee802154/l2/src/ieee802154_test.c [hunk 3]
```diff
@@ -177,6 +177,42 @@ static int tear_down_short_addr(struct net_if *iface, struct ieee802154_context
...
+	}
+
+	return true;
+}
+
+static void tear_down_security(void)
+{
+	struct ieee802154_context *ctx = net_if_l2_data(iface);
+
+	if (ctx->sec_ctx.level == IEEE802154_SECURITY_LEVEL_NONE)
```
[the remark this answers is not in the document]
Comment (fgrandel) [reply tests/net/ieee802154/l2/src/ieee802154_test.c]: Fixed

File: tests/net/ieee802154/l2/src/ieee802154_test.c (hunk 1 above)
Comment (fgrandel) [reply tests/net/ieee802154/l2/src/ieee802154_test.c:1146]: Gotcha, will do.

Comment (fgrandel) [reply tests/net/ieee802154/l2/src/ieee802154_test.c]: Fixed.

File: tests/net/ieee802154/l2/src/ieee802154_test.c [hunk 4]
```diff
@@ -177,6 +177,44 @@ static int tear_down_short_addr(struct net_if *iface, struct ieee802154_context
...
+{
+	struct ieee802154_context *ctx = net_if_l2_data(iface);
+
+	if (ctx->sec_ctx.level == IEEE802154_SECURITY_LEVEL_NONE) {
+		return;
+	}
+
+	cipher_free_session(ctx->sec_ctx.enc.device, &ctx->sec_ctx.enc);
+	cipher_free_session(ctx->sec_ctx.dec.device, &ctx->sec_ctx.dec);
+	ctx->sec_ctx.level = IEEE802154_SECURITY_LEVEL_NONE;
```
Comment (nandojve) [inline tests/net/ieee802154/l2/src/ieee802154_test.c:304]: Did you know why it is necessary clean-up using cipher* functions instead make a request passing security_level as IEEE802154_SECURITY_LEVEL_NONE ? This seems to be a little hack to me.

**Review by nandojve — DISMISSED** [review]
Hi @fgrandel ,

I only noted that when you wrote the tear down seems that expose something little odd to me.

Thank you for heads up and these changes.

File: tests/net/ieee802154/l2/src/ieee802154_test.c (hunk 4 above)
Comment (fgrandel) [reply tests/net/ieee802154/l2/src/ieee802154_test.c]: This is a very good question. The fact that the security context is not cleaned up inside the mgmt-stack before starting a new one is even a resource leak. Fixed.

**Review by nandojve — APPROVED** [review]

**Review by jukkar — APPROVED** [review]
LGTM

**Review by rlubos — APPROVED** [review]

**Pull request merged**